SimpleExec/ProcessExtensions.cs

```diff
@@ -12,12 +14,31 @@ public static void Run(this Process process, bool noEcho, string echoPrefix)
             process.WaitForExit();
         }
 
-        public static Task RunAsync(this Process process, bool noEcho, string echoPrefix)
+        public static Task RunAsync(this Process process, bool noEcho, string echoPrefix, CancellationToken cancellationToken = default)
         {
             var tcs = new TaskCompletionSource<object>();
-            process.Exited += (s, e) => tcs.SetResult(default);
+            var cancelled = false;
+
+            process.Exited += (s, e) =>
+            {
+                if (cancelled)
+                {
+                    tcs.TrySetCanceled(cancellationToken);
+                }
+                else
+                {
+                    tcs.SetResult(default);
+                }
+            };
+
             process.EnableRaisingEvents = true;
             process.EchoAndStart(noEcho, echoPrefix);
+
+            cancellationToken.Register(() => {
+                process.Kill();

```

Comment:
This can throw in number of different ways, the most common of which is when the normal process termination coincides with cancellation.

Reply:
Is that good or bad?

Reply:
It can throw an exception inside the cancellation registration, which will in turn propagate to where the cancellation was requested (in case it was triggered with `cts.Cancel()`) or to the timer thread (in case it was triggered with `cts.CancelAfter()`). You basically have two options:

1. Swallow all exceptions and just hope it successfully kills the process. Make sure you release the TCS immediately thought, because in case of failure it will result in a deadlock.
2. Swallow and don't release TCS, but then spawn a delayed killswitch that breaks out the TCS out of deadlock in case the termination has failed.

Option 1 is the easiest, but you won't be able to guarantee that the process gets killed. Also, the method will often return before the process terminates, even in case of success.

Option 2 is harder, but more reliable. The method will not return until the process stopped breathing or until the termination has timed out. This is the approach that [CliWrap](https://github.com/Tyrrrz/CliWrap) uses.

Hitting exceptions with `process.Kill()` is practically guaranteed in the long run, especially with small and fast-running processes.

Reply:
Could you show me the code for option 2?

Reply:
https://github.com/Tyrrrz/CliWrap/blob/7c88bcb22d0e61cecfe3c3981622e3a3993f1164/CliWrap/Command.cs#L386-L394

Reply:
This is the code for option 2 basically:

https://github.com/Tyrrrz/CliWrap/blob/7c88bcb22d0e61cecfe3c3981622e3a3993f1164/CliWrap/Internal/ProcessEx.cs#L84-L104

The important part is that, if an exception happened, it waits 3 seconds and cancels the process task completion source.

The part that you linked is related to piping, which this project doesn't seem to work with, so it's not necessary.

Reply:
Looking at that code, doesn't that mean code hangs for 3 seconds every time a process is canceled?

Reply:
I decided to go with Option 1, and if for someone reason Option 2 becomes necessary for devs, they should probably look to Option 3... praying.